service/src/tests/mocks/parentchain_api_mock.rs

```diff
@@ -0,0 +1,72 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use itp_api_client_extensions::{ApiResult, ChainApi, StorageProof};
+use itp_test::builders::{
+	parentchain_block_builder::ParentchainBlockBuilder,
+	parentchain_header_builder::ParentchainHeaderBuilder,
+};
+use itp_types::{SignedBlock, H256};
+use sp_finality_grandpa::AuthorityList;
+use std::cmp::min;
+
+pub struct ParentchainApiMock {
+	parentchain: Vec<SignedBlock>,
+}
+
+impl ParentchainApiMock {
+	pub(crate) fn new() -> Self {
+		ParentchainApiMock { parentchain: Vec::new() }
+	}
+
+	// can this be combined with the constructor? Does it make sense?
+	pub fn create_mock_parentchain(&mut self, len: u32) {
+		for n in 1..=len {
+			let header = ParentchainHeaderBuilder::default().with_number(n).build();
+			let block = ParentchainBlockBuilder::default().with_header(header).build_signed();
+			self.parentchain.push(block);
+		}
+	}
+}
+
+impl ChainApi for ParentchainApiMock {
+	fn last_finalized_block(&self) -> ApiResult<Option<SignedBlock>> {
+		let i = self.parentchain.len();
+		ApiResult::Ok(Some(self.parentchain[i - 1].clone()))
```

Comment:
use `last()` instead of working with indices (using indices should be a rare exception)

Reply:
Yes, however, indexing with safety checks might be useful to get owned values without clones.

Reply:
You can easily get owned values without indices as well ☺️

Reply:
Just for documentation's sake (Gaudeny and I discussed this in Teams):
```
fn last_finalized_block(&self) -> ApiResult<Option<SignedBlock>> {
    ApiResult::Ok(self.parentchain.last().cloned())
}
```

Reply:
But this is with clone? In this case, a clone is necessary anyhow, so it does not matter.

Reply:
We need a clone here, yes. In case you don't want an owned value, you can of course always just return a reference (and drop the `cloned()`.

Reply:
Yes, but what I was trying to say: If you own the vector already, you can extract an owned value by indexing, and don't need to clone, which might be useful sometimes.

```
let vec: Vec<MySuperComplexStruct>= vec![complex_struct];

if !vec.is_empty() {
     let cs: MySuperComplexStruct = vec[0]
}
```

Reply:
So you're saying the `[]`-operator moves? And in your case the first element will be moved and the vector will be unusable afterwards?
When I wrote this test code, it did not compile, saying that you need the copy trait on your struct, suggesting that `[]` does not really move, but creates a copy?

Reply:
(We should probably discuss this in person, to not spam this PR. But I'm interested to learn more 😄 )